src/quadiron_c.cpp

```diff
@@ -0,0 +1,388 @@
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    int need_decode = 0;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        /*
+         * Easy case where the target is a data then simply decode
+         */
+        if (destination_idx < fec->n_data) {
+            wanted_idxs_vec[destination_idx] = true;
+            res = fec->decode_blocks_vertical(
+                data_vec,
+                parities_vec,
+                parities_props,
+                missing_idxs_vec,
+                wanted_idxs_vec,
+                block_size);
+            if (!res) {
+                return -1;
+            }
+            return 0;
+        }
+    }
+
+    /*
+     * At this point we want a parity to be reconstructed
+     * If systematic we may need to decode if a data is missing.
+     * If non-systematic we always need to decode
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (missing_idxs[i]) {
+                need_decode = 1;
+                wanted_data_vec[i] = true;
+                std::vector<uint8_t>* block =
+                    new std::vector<uint8_t>(block_size);
+                data_vec[i] = block->data();
+            }
+        }
+    } else {
+        need_decode = 1;
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            wanted_data_vec[i] = true;
+            std::vector<uint8_t>* block = new std::vector<uint8_t>(block_size);
+            data_vec[i] = block->data();
+        }
+    }
+
+    if (need_decode) {
+        res = fec->decode_blocks_vertical(
+            data_vec,
+            parities_vec,
+            parities_props,
+            missing_idxs_vec,
+            wanted_data_vec,
+            block_size);
+        if (!res) {
+            return -1;
+        }
+    }
+
+    /*
+     * At this point we have all data blocks.
+     */
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        wanted_idxs_vec[destination_idx - fec->n_data] = true;
+    } else {
+        wanted_idxs_vec[destination_idx] = true;
+    }
+
+    fec->encode_blocks_vertical(
+        data_vec, parities_vec, parities_props, wanted_idxs_vec, block_size);
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (i == destination_idx - fec->n_data) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[i].fnt_serialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1) {
+                    return -1;
```

Comment:
this code leaks because we don't delete allocated buffers @slaperche-scality and apparently block_size is too big for the stack. do you have any suggestion except goto ?

Reply:
OK thanks, I will try

Reply:
Okay this was a scope issue: the variable was released at end of code block

Reply:
```
==13817== 
==13817== HEAP SUMMARY:
==13817==     in use at exit: 85,512 bytes in 23 blocks
==13817==   total heap usage: 413,765 allocs, 413,742 frees, 4,483,509,795 bytes allocated
==13817== 
==13817== LEAK SUMMARY:
==13817==    definitely lost: 0 bytes in 0 blocks
==13817==    indirectly lost: 0 bytes in 0 blocks
==13817==      possibly lost: 0 bytes in 0 blocks
==13817==    still reachable: 85,512 bytes in 23 blocks
==13817==         suppressed: 0 bytes in 0 blocks
```